### PR TITLE
Add gce disk labels support via create volume parameters

### LINKED ISSUE #340: Support for disk labels parameter
# Problem

I use PVC and PV resources on GKE, and I also use PersistentDisk(PD) for ComputeEngine instances.
I usually filter PDs by labels when I delete it.
However, the PD which created by PV does not have labels for filtering.
This makes me difficult to filter to delete PDs.

# Solution

Adding parameter to StorageClass and use it in this driver.

```
https://kubernetes.io/docs/concepts/storage/storage-classes/#gce-pd
e.g.

apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: slow
provisioner: kubernetes.io/gce-pd
parameters:
  type: pd-standard
  replication-type: none
  disk-labels: 'foo=123,bar=456' # Add this parameter to set labels
```

# Additional Info

I read the related issue on [external-provisioner](https://github.com/kubernetes-csi/external-provisioner/issues/86), but in my use case, I think it's better to implement on this driver, since it depends on GCP PD.

Sample implementation:
https://github.com/katsew/gcp-compute-persistent-disk-csi-driver/tree/disk-labels-support/pkg

### PR DESCRIPTION
**Pull request by mattcary**
Labels are validated to match gce requirements.

Cherry-picked from #566

**Which issue(s) this PR fixes**:
Fixes #340

```release-note
Disk labels support via CreateVolume (and hence StorageClass) parameters
```

### CODE REVIEW

File: test/e2e/tests/single_zone_e2e_test.go [hunk 1]
```diff
@@ -292,6 +292,46 @@ var _ = Describe("GCE PD CSI Driver", func() {
...
+			"key2": "value2",
+			// The label below is added as an --extra-label driver command line argument.
+			testutils.DiskLabelKey: testutils.DiskLabelValue,
+		}))
+		Expect(cloudDisk.Name).To(Equal(volName))
+
+		defer func() {
+			// Delete Disk
+			client.DeleteVolume(volID)
+			Expect(err).To(BeNil(), "DeleteVolume failed")
```
Comment (Jiawei0227) [inline test/e2e/tests/single_zone_e2e_test.go:327]: Which err?

File: test/k8s-integration/config/sc-standard.yaml [hunk 2]
```diff
@@ -5,4 +5,5 @@ metadata:
 provisioner: pd.csi.storage.gke.io
 parameters:
   type: pd-standard
+  labels: key1=value1,key2=value2
```
Comment (Jiawei0227) [inline test/k8s-integration/config/sc-standard.yaml:8]: Maybe add a comment here saying this is for tests?

File: test/e2e/tests/single_zone_e2e_test.go (hunk 1 above)
Comment (mattcary) [reply test/e2e/tests/single_zone_e2e_test.go:327]: Good catch, thanks.

File: test/k8s-integration/config/sc-standard.yaml (hunk 2 above)
Comment (mattcary) [reply test/k8s-integration/config/sc-standard.yaml:8]: Good idea, done.

File: pkg/common/parameters_test.go [hunk 3]
```diff
@@ -37,38 +37,41 @@ func TestExtractAndDefaultParameters(t *testing.T) {
...
 				ReplicationType:      "none",
 				DiskEncryptionKMSKey: "",
-				Tags:                 make(map[string]string),
-				Labels:               make(map[string]string),
+				Tags:                 map[string]string{},
+				Labels:               map[string]string{},
 			},
 		},
 		{
 			name:       "random keys",
```
Comment (saikat-royc) [inline pkg/common/parameters_test.go:57]: is this tc name correct?

File: pkg/common/parameters_test.go [hunk 4]
```diff
@@ -37,38 +37,41 @@ func TestExtractAndDefaultParameters(t *testing.T) {
...
-			parameters: map[string]string{ParameterKeyType: "", "foo": "", ParameterKeyDiskEncryptionKmsKey: ""},
+			parameters: map[string]string{ParameterKeyType: "", "foo": "", ParameterKeyDiskEncryptionKmsKey: "", ParameterKeyLabels: ""},
 			labels:     map[string]string{},
 			expectErr:  true,
 		},
 		{
 			name:       "real values",
```
Comment (saikat-royc) [inline pkg/common/parameters_test.go:63]: maybe we should elaborate the tc name, as labels picked from params

File: pkg/common/utils.go [hunk 5]
```diff
@@ -156,3 +157,60 @@ func CreateNodeID(project, zone, name string) string {
...
+func ConvertLabelsStringToMap(labels string) (map[string]string, error) {
+	const labelsDelimiter = ","
+	const labelsKeyValueDelimiter = "="
+
+	labelsMap := make(map[string]string)
+	if labels == "" {
+		return labelsMap, nil
+	}
+
+	regexKey, _ := regexp.Compile(`^\p{Ll}[\p{Ll}0-9_-]{0,62}$`)
```
Comment (saikat-royc) [inline pkg/common/utils.go:172]: Maybe we can put a pointer to the label format https://cloud.google.com/compute/docs/labeling-resources#label_format

File: examples/kubernetes/demo-labeled-sc.yaml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: csi-gce-pd
+provisioner: pd.csi.storage.gke.io
+parameters:
```
[the remark this answers is not in the document]
Comment (mattcary) [reply examples/kubernetes/demo-labeled-sc.yaml:8]: done

File: pkg/common/parameters_test.go (hunk 4 above)
Comment (mattcary) [reply pkg/common/parameters_test.go]: done

File: pkg/common/utils.go (hunk 5 above)
Comment (mattcary) [reply pkg/common/utils.go:173]: done

**Pull request merged**